zio-http/shared/src/main/scala/zio/http/URL.scala

```diff
@@ -281,24 +282,31 @@ final case class URL(
 object URL {
   def empty: URL = URL(Path.empty)
 
-  def decode(string: String): Either[Exception, URL] = {
-    def invalidURL(string: String) = Left(new MalformedURLException(s"""Invalid URL: "$string""""))
+  /**
+   * Proper implementation of [[java.net.MalformedURLException]] which, unlike
+   * the Java version, can propagate the cause.
+   */
+  final case class MalformedURLException(message: String, cause: Option[Throwable])
+      extends IOException(message, cause.orNull)
```

Comment:
I think we can change this so that we avoid building the string for cases that we discard the error:

```scala
  final case class MalformedURLException(url: String, cause: Option[Throwable]) extends IOException(cause.orNull) {
    override def getMessage: String = s"""Invalid URL: "$url""""
  }
```

Reply:
Changed both. Thanks